### PR TITLE
Fuzz multiple targets in cranelift-icache

### PR DESCRIPTION
**Pull request by elliottt**
Fuzz additional targets in the cranelift-icache target. The list of targets fuzzed is controlled by the targets enabled in `fuzz/Cargo.toml`.

This PR also reworks how instruction disabling is done in function generator, moving the deny-list to a function to make the decision at runtime instead of compile time.

### CODE REVIEW

File: cranelift/fuzzgen/src/lib.rs
Comment (afonso360) [inline cranelift/fuzzgen/src/lib.rs:56]: I can't comment on the exact line (61), but can we print only the correct `target {}` line here?

File: fuzz/fuzz_targets/cranelift-compile.rs
Comment (afonso360) [inline fuzz/fuzz_targets/cranelift-compile.rs:55]: I think we might be able to reuse this for the icache fuzzer as well. Feel free to ignore it if it's too much work to change that.

**Review by afonso360 — COMMENTED** [review]
I like this! It gets us some more coverage without needing native hosts.

We are probably going to run into a bunch of unimplemented instructions in s390x and riscv64 which haven't been fuzzed by fuzzgen before.

Another issue is that right now we `cfg` out unimplemented instructions in [function_generator.rs](https://github.com/bytecodealliance/wasmtime/blob/b47e644c3dcd857221ef7163e780a0399a183966/cranelift/fuzzgen/src/function_generator.rs#L251-L260) which is not going to play well with this!

I think we could change that to have a global array with all the opcode signatures like we do now, and filter them in a separate function based on the arch provided.

File: fuzz/fuzz_targets/cranelift-compile.rs
Comment (elliottt) [reply fuzz/fuzz_targets/cranelift-compile.rs:55]: That's a great point! I'd be happy to pull this type back into cranelift-fuzzgen so that we can use it in both targets :+1:

File: cranelift/fuzzgen/src/lib.rs
Comment (elliottt) [reply cranelift/fuzzgen/src/lib.rs:56]: Yep! I'll duplicate the match in cranelift-compile.rs here.

File: cranelift/fuzzgen/src/lib.rs
Comment (elliottt) [inline cranelift/fuzzgen/src/lib.rs:67]: @afonso360 is this what you had in mind?

File: fuzz/fuzz_targets/cranelift-compile.rs
Comment (elliottt) [reply fuzz/fuzz_targets/cranelift-compile.rs]: I've added a new struct `FunctionWithIsa` in cranelift-fuzzgen, and will rework cranelift-icache in a separate PR once this one has landed.

File: cranelift/fuzzgen/src/lib.rs
Comment (afonso360) [reply cranelift/fuzzgen/src/lib.rs:127]: Yeah that looks good!

The reason for that is mostly that when we generate a function it is only guaranteed to compile for a particular architecture (due to unimplemented instructions), and having all of them there is slightly misleading.

File: cranelift/fuzzgen/src/lib.rs
Comment (jameysharp) [inline cranelift/fuzzgen/src/lib.rs:294]: `run_func_passes` builds a `TargetIsa` which probably doesn't match `self.isa`. I think the passes it runs probably don't actually look at anything except the target-independent flags, so that's probably fine, but I'm not sure. I think I've been confused about this before and forgotten the answer. Maybe @afonso360 knows?

**Review by jameysharp — APPROVED** [review]
This all looks right! (I think: I'm not completely sure about `run_func_passes`.)

I discussed this with Trevor and we decided not to merge this until January, when people will be back from holidays.

I'd like it if we put a little more thought into making `cranelift-icache` test various targets and flags, rather than adding a new fuzz target. I don't know for sure that it's a good idea to combine them, but I don't see a strong reason to keep them separate. And as I understand it, we have a fixed CPU budget on OSS-Fuzz to split between all our fuzz targets, so there's some cost to adding new ones. That said, I'm okay with adding the separate fuzz target now and combining them later.

Trevor tells me that he ran this fuzz target for four hours without finding any bugs, which actually has me concerned that it may not be testing what we wanted to test. :sweat_smile: I'm hoping output from `cargo fuzz coverage` will be revealing, but maybe we should also try removing a lowering from a backend and making sure that the fuzz target finds that...

**Comment by afonso360** [comment]
>Trevor tells me that he ran this fuzz target for four hours without finding any bugs, which actually has me concerned that it may not be testing what we wanted to test. 

That's weird, when I ran this it crashed almost immediately even without a previous corpus.

Ran it again with (This is just the run string that I usually use for fuzzgen):
`cargo +nightly fuzz run --no-default-features cranelift-compile -- -rss_limit_mb=8192 -max_len=128 -runs=5000000`

Gives me the following input (base64'd):
```
////L/////////////////8K/wAAXQAAAAD/AQAAAA==
```

Crashes with:
```
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Undeclared function u0:0 is referenced by u1:0!', fuzz/fuzz_targets/cranelift-compile.rs:60:53
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

File: cranelift/fuzzgen/src/lib.rs
Comment (afonso360) [reply cranelift/fuzzgen/src/lib.rs:294]: Yeah it does. We only need to build it because `canonicalize_nans` requires it, although it doesn't use it besides the `enable_verifier` flag.

It would remove quite a bit of code to not have to build it again, but it would be nice if we could avoid verifying it there, since I benchmarked it once and it ended up showing on the profiling. (We still run the verifier while compiling)

The other passes don't use the target ISA.

**Comment by afonso360** [comment]
A follow up from the issue above, the default fuzzgen config generates random function references even if they don't actually exist. This works for icache, but [not for fuzzgen](https://github.com/bytecodealliance/wasmtime/blob/fac4a915a3848f7efc26bdc50b6545b476545d35/cranelift/fuzzgen/src/lib.rs#L311-L314) since there we actually try to do some linking (via cranelift-jit).

Altering that config for this target does get it past that error.

**Comment by jameysharp** [comment]
Huh, I guess I'm surprised that `cranelift-jit` works at all for cross-compiling. I don't think we should try to test those code paths when cross-compiling. So I think this is another argument for merging this new fuzz target into `cranelift-icache`, or at least making it more closely resemble that fuzz target.

**Comment by alexcrichton** [comment]
Could this perhaps get added as a mode to `cranelift-fuzzgen` instead of adding a new fuzz-target? Mostly due to how oss-fuzz allocates time I think it would be best to not expand the time allocated for Cranelift-specific fuzzing since the time alloted to cranelift-fuzzgen is probably good enough, although I could see this going both ways.

My thinking is that the first byte of the input, for example, and its lowest bit would indicate "run in an interpreter" or "compile, possibly a cross-compile"

**Comment by afonso360** [comment]
Merging this into icache would have the same effect right? We would keep the same number of targets as we do today (i.e. `fuzzgen` and `icache`). I think that would be the best way to go.

**Comment by alexcrichton** [comment]
Oh dear sorry disregard me, indeed!

File: cranelift/fuzzgen/src/function_generator.rs
Comment (jameysharp) [inline cranelift/fuzzgen/src/function_generator.rs:152]: Let's add a comment here like the one in `generate_instructions`:
```suggestion
        // We filter out condition codes that aren't supported by the target at
        // this point after randomly choosing one, instead of randomly choosing a
        // supported one, to avoid invalidating the corpus when these get implemented.
        if matches!(fgen.target_triple.architecture, Architecture::Aarch64(_))
```

File: cranelift/fuzzgen/src/lib.rs
Comment (jameysharp) [inline cranelift/fuzzgen/src/lib.rs:68]: Let's add a comment here like the one in `generate_instructions`:
```suggestion
        // We filter out targets that aren't supported in the current build
        // configuration after randomly choosing one, instead of randomly choosing
        // a supported one, so that the same fuzz input works across different build
        // configurations.
        let target = u.choose(isa::ALL_ARCHITECTURES)?;
```

File: cranelift/fuzzgen/src/function_generator.rs
Comment (jameysharp) [inline cranelift/fuzzgen/src/function_generator.rs:468]: Unimportant and pre-existing nit: I believe the `'static` lifetime isn't required here, so I'd leave it out.
```suggestion
const OPCODE_SIGNATURES: &[OpcodeSignature] = &[
```

File: cranelift/fuzzgen/src/lib.rs
Comment (jameysharp) [inline cranelift/fuzzgen/src/lib.rs:276]: I think this doesn't need to be `pub`; it looks like it's only called from this module now.

File: cranelift/fuzzgen/src/lib.rs
Comment (jameysharp) [inline cranelift/fuzzgen/src/lib.rs:283]: I think this doesn't need to be `pub`; it looks like it's only called from this module now.

**Review by jameysharp — COMMENTED** [review]
I am super excited to see this coming together! I have a few small finishing-touches suggestions but I think this is nearly ready to go.

**Review by afonso360 — APPROVED** [review]
This looks great! Thank you! 🎉

**Review by jameysharp — APPROVED** [review]
I'm delighted. Ship it!

**Pull request merged**
